### PR TITLE
fix: update covariance matrix calculations in metrics.py

### PR DESCRIPTION
**Pull request by Scherlac**
fix: update covariance matrix calculations in metrics.py to include dynamic scaling factors to mitigate isotropic bounding boxes.

Related issue: #23011 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updates OBB ProbIoU covariance computation by introducing per-call scaling of width/height terms before building the covariance matrix. 🎯

### 📊 Key Changes
-Adds `random` import and generates random scaling factors (`pa`, `pb`) inside `probiou()` and `batch_probiou()` 🎲
-Changes `_get_covariance_matrix()` signature to accept scaling parameters and applies them to the `a` and `b` terms (derived from box width/height) 🧩
-Propagates the new parameters through all internal calls to `_get_covariance_matrix()` for consistent scaling between compared OBBs 🔁

### 🎯 Purpose & Impact
-Intended to improve numerical stability of covariance-related terms for OBB ProbIoU calculations 🛠️
-However, introducing randomness in metric computation can make results non-deterministic across runs (even with identical inputs), which may affect reproducibility, evaluation consistency, and debugging reliability ⚠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Scherlac, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist (this is automated, and an engineer will assist shortly) 🤖

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/metrics.py [hunk 1]
```diff
@@ -226,8 +230,10 @@ def probiou(obb1: torch.Tensor, obb2: torch.Tensor, CIoU: bool = False, eps: flo
     """
     x1, y1 = obb1[..., :2].split(1, dim=-1)
     x2, y2 = obb2[..., :2].split(1, dim=-1)
-    a1, b1, c1 = _get_covariance_matrix(obb1)
-    a2, b2, c2 = _get_covariance_matrix(obb2)
+    pa = random.uniform(1.0, 2.0)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: ❗ **CRITICAL**: Introducing `random.uniform()` inside `probiou()` makes this metric non-deterministic across runs for identical inputs, which can break evaluation reproducibility and make regressions hard to debug. If scaling is needed for numerical stability, it should be deterministic (e.g., a constant factor, or a factor derived from the input tensors), or exposed as an explicit function argument so callers control it.

File: ultralytics/utils/metrics.py [hunk 2]
```diff
@@ -268,10 +274,13 @@ def batch_probiou(obb1: torch.Tensor | np.ndarray, obb2: torch.Tensor | np.ndarr
     obb1 = torch.from_numpy(obb1) if isinstance(obb1, np.ndarray) else obb1
     obb2 = torch.from_numpy(obb2) if isinstance(obb2, np.ndarray) else obb2
 
+    pa = random.uniform(1.0, 2.0)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: ❗ **CRITICAL**: Same issue as `probiou()`: randomness in `batch_probiou()` will make batch evaluation non-reproducible and can cause flaky tests/benchmarks. Consider a deterministic scaling strategy or a caller-provided parameter instead of sampling per call.

File: ultralytics/utils/metrics.py [hunk 3]
```diff
@@ -187,7 +188,7 @@ def kpt_iou(
     return ((-e).exp() * kpt_mask[:, None]).sum(-1) / (kpt_mask.sum(-1)[:, None] + eps)
 
 
-def _get_covariance_matrix(boxes: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+def _get_covariance_matrix(boxes: torch.Tensor, pa, pb) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: ⚠️ **HIGH**: Changing `_get_covariance_matrix()` to require `pa, pb` risks breaking any other internal call sites (in this file or elsewhere) that aren’t updated in the same PR. Providing defaults preserves backward compatibility and makes the function safer to use (e.g., default scaling of 1.0 keeps current behavior when scaling isn’t needed).

**Suggested change:**
```suggestion
def _get_covariance_matrix(boxes: torch.Tensor, pa: float = 1.0, pb: float = 1.0) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```

File: ultralytics/utils/metrics.py [hunk 4]
```diff
@@ -199,6 +200,9 @@ def _get_covariance_matrix(boxes: torch.Tensor) -> tuple[torch.Tensor, torch.Ten
     # Gaussian bounding boxes, ignore the center points (the first two columns) because they are not needed here.
     gbbs = torch.cat((boxes[:, 2:4].pow(2) / 12, boxes[:, 4:]), dim=-1)
     a, b, c = gbbs.split(1, dim=-1)
+    # a = 10.0 * a  # scale width and height by 10 for stability
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: 📝 **LOW**: There’s commented-out scaling logic left in the function. Since this code path is performance-sensitive and widely used, keeping dead/commented code here adds noise and makes it harder to reason about the intended scaling behavior. Recommend removing it once the chosen approach is settled.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Main concern: adding per-call randomness to ProbIoU makes metrics non-deterministic, which is a serious regression for evaluation reproducibility and debugging. Also, `_get_covariance_matrix()` now requires new parameters; adding defaults would reduce risk of breaking other call sites. Minor: remove leftover commented-out scaling line for maintainability.

💬 Posted 4 inline comments

**Comment by Scherlac** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Scherlac** [comment]
@glenn-jocher,
Any suggestion on fixing the randomness eg. using a deterministic seed. Is there some architectural component in ultralytics for that?

**Comment by glenn-jocher** [comment]
Randomness in `probiou()`/`batch_probiou()` isn’t something we can accept for evaluation metrics (even if you seed it), since it makes mAP/PR curves input-order/worker dependent and harder to reproduce/debug; instead, please make the covariance stabilization deterministic (e.g., clamp `w/h` away from 0, add a small fixed diagonal “jitter” to the `a/b` terms, or regularize the determinant path) and add a small regression test around the isotropic-case from #23011 in `ultralytics/utils/metrics.py` (see the current implementations of [`probiou()` and `_get_covariance_matrix()`](https://docs.ultralytics.com/reference/utils/metrics/)). For completeness, we do have a global seeding helper for training runs via [`init_seeds()`](https://docs.ultralytics.com/reference/utils/torch_utils/#ultralyticsutils_torch_utilsinit_seeds), but it’s not intended to be called inside metrics code.

**Comment by Scherlac** [comment]
@glenn-jocher I've implemented a bit different approach with individual scale factors for affected bboxes. It seems to pass benchmark and performs well on my yolo-shapes dataset. The implementation might raise some concerns I guess :D  
I will try to come up with regression test.

**Comment by glenn-jocher** [comment]
That approach can work as long as the “per-bbox scale factors” are fully deterministic (purely derived from the box values like `w/h`, no RNG/global state) and applied symmetrically so `batch_probiou(a, b)` is consistent regardless of call order; once you add the regression test we can review the tradeoff (the [model testing guide](https://docs.ultralytics.com/guides/model-testing/) shows the general pattern we use).

```python
import torch
from ultralytics.utils.metrics import batch_probiou

def test_batch_probiou_isotropic_stable():
    # near/isotropic OBBs (xywhr, radians)
    gt = torch.tensor([[50.0, 50.0, 40.0, 40.0, 0.7853982]], dtype=torch.float32)
    pr = torch.tensor([[50.0, 50.0, 40.0, 40.0, 0.7853982]], dtype=torch.float32)

    iou1 = batch_probiou(gt, pr)
    iou2 = batch_probiou(gt, pr)

    assert torch.allclose(iou1, iou2)
    assert torch.isfinite(iou1).all()
    assert (iou1 >= 0).all() and (iou1 <= 1).all()
```

**Comment by Scherlac** [comment]
@glenn-jocher 

>  add the regression test we can review the tradeoff

Can you please point to an existing regression test for a specific feature, it might help me to understand the scope.

I`ve tested the fix on my initial use case already back then: [see](https://github.com/Scherlac/yolo-shapes/tree/main/runs/obb/yolo_obb_shapes_training-fix-individual-scale)

**Comment by glenn-jocher** [comment]
Yep—our best “feature/regression test” patterns are the small pytest smoke tests in the repo’s [tests/ folder](https://github.com/ultralytics/ultralytics/tree/main/tests), especially [tests/test_python.py](https://github.com/ultralytics/ultralytics/blob/main/tests/test_python.py) (Python API) and [tests/test_cli.py](https://github.com/ultralytics/ultralytics/blob/main/tests/test_cli.py) / [tests/test_exports.py](https://github.com/ultralytics/ultralytics/blob/main/tests/test_exports.py) (CLI/export); for your `batch_probiou` isotropic fix, I’d add a new fast unit test alongside those (e.g. `tests/test_metrics.py`) that uses only synthetic `torch.tensor` OBBs and asserts finiteness + determinism + symmetry (swap `obb1/obb2` gives the transpose) so it’s stable in CI without any dataset downloads.
